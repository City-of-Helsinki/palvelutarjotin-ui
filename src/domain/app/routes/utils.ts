import { CollectionItemType } from 'react-helsinki-headless-cms';

import { ROUTES } from './constants';

export const getCmsPagePath = (uri?: string | null): string => {
  return ROUTES.CMS_PAGE.replace('/:uri', uri ?? '');
};

export const getCmsArticlePath = (uri?: string | null): string => {
  return ROUTES.CMS_ARTICLE.replace('/:uri', uri ?? '');
};

export const getCollectionItemUrl = (item: CollectionItemType): string => {
  if (!item) {
    return '#';
  }
  if (item.__typename === 'Post') {
    return getCmsArticlePath(item.uri);
  }
  if (item.__typename === 'Page') {
    return getCmsPagePath(item.uri);
  }
  return item?.uri ?? '';
};
