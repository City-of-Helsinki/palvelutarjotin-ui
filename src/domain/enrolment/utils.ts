import {
  EnrolOccurrenceMutationInput,
  Language,
  NotificationType,
} from '../../generated/graphql';
import { EnrolmentFormFields } from './enrolmentForm/constants';

const getNotificationType = (values: EnrolmentFormFields): NotificationType => {
  return values.hasEmailNotification && values.hasSmsNotification
    ? NotificationType.EmailSms
    : NotificationType.Email;
};

export const getEnrolmentPayload = ({
  occurrenceIds,
  values,
}: {
  occurrenceIds: string[];
  values: EnrolmentFormFields;
}): EnrolOccurrenceMutationInput => {
  return {
    occurrenceIds,
    notificationType: getNotificationType(values),
    person: values.isSameResponsiblePerson
      ? undefined
      : { ...values.person, language: values.language as Language },
    studyGroup: {
      ...values.studyGroup,
      amountOfAdult: Number(values.studyGroup.amountOfAdult),
      groupSize: Number(values.studyGroup.groupSize),
      person: {
        ...values.studyGroup.person,
        language: values.language as Language,
      },
      studyLevels: values.studyGroup.studyLevels,
    },
  };
};

declare let grecaptcha: ReCAPTCHA;

interface ReCAPTCHA {
  ready: (f: () => void) => void;
  execute: (key: string, action: { action: string }) => Promise<string>;
}

export const getCAPTCHAToken = (): Promise<string> => {
  return new Promise((resolve, reject) => {
    if (process.env.NEXT_PUBLIC_CAPTCHA_KEY) {
      const captchaKey = process.env.NEXT_PUBLIC_CAPTCHA_KEY;
      grecaptcha.ready(() => {
        grecaptcha
          .execute(captchaKey, { action: 'submit' })
          .then(resolve)
          .catch(reject);
      });
    } else {
      reject('CAPTCHA_KEY missing!');
    }
  });
};
