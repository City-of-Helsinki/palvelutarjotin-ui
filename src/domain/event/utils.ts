import { isToday, isTomorrow } from 'date-fns';
import { TFunction } from 'next-i18next';

import { EventFieldsFragment } from '../../generated/graphql';
import { Language } from '../../types';
import formatDate from '../../utils/formatDate';
import getLocalisedString from '../../utils/getLocalisedString';
import getTimeFormat from '../../utils/getTimeFormat';
import { EVENT_PLACEHOLDER_IMAGES } from './constants';

/**
 * Get event placeholder image url
 * @param {string} id
 * @return {string}
 */
export const getEventPlaceholderImage = (id: string): string => {
  const numbers = id.match(/\d+/g);
  const sum = numbers
    ? numbers.reduce((prev: number, cur: string) => prev + Number(cur), 0)
    : 0;
  const index = sum % 4;

  return EVENT_PLACEHOLDER_IMAGES[index];
};

export const getEventStartTimeStr = (
  event: EventFieldsFragment,
  locale: Language,
  t: TFunction
): string | null => {
  const startTime = event.startTime ? new Date(event.startTime) : null;
  const timeFormat = getTimeFormat(locale);

  if (!startTime) return null;

  if (isToday(startTime))
    return t('event:eventCard.startTime.today', {
      time: formatDate(startTime, timeFormat, locale),
    });

  if (isTomorrow(startTime))
    return t('event:eventCard.startTime.tomorrow', {
      time: formatDate(startTime, timeFormat, locale),
    });

  return t('event:eventCard.startTime.other', {
    date: formatDate(startTime),
    time: formatDate(startTime, timeFormat, locale),
  });
};

/**
 * Check is event free
 * @param eventData
 * @return {boolean}
 */
export const isEventFree = (event: EventFieldsFragment): boolean =>
  Boolean(event.offers.find((item) => item.isFree)?.isFree);

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const getEventFields = (
  event: EventFieldsFragment | undefined | null,
  locale: Language
) =>
  event
    ? {
        isEventFree: isEventFree(event),
        eventName: getLocalisedString(event?.name, locale),
        shortDescription: getLocalisedString(event?.shortDescription, locale),
        description: getLocalisedString(event?.description, locale),
        imageUrl: event?.images?.[0]?.url,
        imageAltText: event?.images?.[0]?.altText,
        photographerName: event?.images?.[0]?.photographerName,
        organization: event.pEvent?.organisation?.name,
        contactPhoneNumber: event.pEvent?.contactPhoneNumber,
        contactEmail: event.pEvent?.contactEmail,
        contactPerson: event.pEvent?.contactPerson?.name,
      }
    : {};
