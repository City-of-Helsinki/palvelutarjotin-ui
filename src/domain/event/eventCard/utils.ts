import { TFunction } from 'next-i18next';

import { EnrolmentStatusKeywordPropsType } from '../../../common/components/keyword/EnrolmentStatusKeyword';
import { EventsFieldsFragment } from '../../../generated/graphql';
import { Language } from '../../../types';
import { formatDateRange } from '../../../utils/time/format';
import { isMultidayOccurrence } from '../../occurrence/utils';
import { getEventStartTimeStr } from '../utils';

export const getOccurrenceDateStr = (
  occurrence: { startTime: string; endTime: string },
  locale: Language,
  t: TFunction
): string | null => {
  if (isMultidayOccurrence(occurrence)) {
    return formatDateRange(
      new Date(occurrence.startTime),
      new Date(occurrence.endTime)
    );
  }

  return getEventStartTimeStr(new Date(occurrence.startTime), locale, t);
};

/**
 * Has the event enrolments enabled in any of it's occurrences?
 * @param event The event to check the status for.
 * @returns `true` if any of the event occurrences has enrolments and
 *  `false` if none.
 */
export const hasEventEnrolmentsEnabled = (event: EventsFieldsFragment) =>
  Boolean(event.pEvent.enrolmentStart) || hasEventExternalEnrolments(event);

/**
 * Has the event external enrolments enabled in any of it's occurrences?
 *
 * @param event The event to check the status for.
 * @returns `true` if the event has some external enrolments and
 *  `false` if none of the occurrences has external enrolments.
 */
export const hasEventExternalEnrolments = (event: EventsFieldsFragment) =>
  Boolean(event.pEvent.externalEnrolmentUrl);

/**
 * Resolve the enrolment status of an event. Has the event free spots available?
 *
 * @param event The event to check the status for.
 * @returns Returns `null` if the enrolment status could not be resolved or
 *    the event has no enrolments at all,
 *    `true` if there are any free spots available and `false` if there are no
 *    spots available.
 */
export const hasEventFreeSpotsAvailable = (
  event: EventsFieldsFragment
): boolean | null => event.pEvent.hasSpaceForEnrolments ?? null;

/**
 * Determines the enrollment status of an event.
 *
 * @param event The event to check the enrollment status for.
 * @returns The enrollment status of the event, or `undefined`
 *      if no enrolments exist or the status could not be determined.
 */
export const getEventEnrolmentStatus = (
  event: EventsFieldsFragment
): EnrolmentStatusKeywordPropsType['enrolmentStatus'] => {
  if (hasEventEnrolmentsEnabled(event)) {
    if (hasEventExternalEnrolments(event)) {
      return 'externally_unknown';
    }
    const hasEventFreeSpots = hasEventFreeSpotsAvailable(event);

    if (hasEventFreeSpots === false) {
      // Event has no free spots available, so it is full.
      return 'full';
    } else if (hasEventFreeSpots === true) {
      // Event has some free spots available.
      return 'free_space_available';
    }
  }
  // No enrolments at all or the enrolment status could not be resolved
  return undefined;
};
