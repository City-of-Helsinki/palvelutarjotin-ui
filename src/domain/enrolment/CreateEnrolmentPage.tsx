import { Button, IconArrowLeft, Notification } from 'hds-react';
import omit from 'lodash/omit';
import { useRouter } from 'next/router';
import React from 'react';
import { useTranslation } from 'react-i18next';
import { toast } from 'react-toastify';

import LoadingSpinner from '../../common/components/loadingSpinner/LoadingSpinner';
import {
  useEventQuery,
  useEnrolOccurrenceMutation,
} from '../../generated/graphql';
import useLocale from '../../hooks/useLocale';
import { Router } from '../../i18n';
import { translateValue } from '../../utils/translateUtils';
import Container from '../app/layout/Container';
import PageWrapper from '../app/layout/PageWrapper';
import { ROUTES } from '../app/routes/constants';
import { getEventFields } from '../event/utils';
import NotFoundPage from '../notFoundPage/NotFoundPage';
import {
  hasOccurrenceSpace,
  isEnrolmentClosed,
  isEnrolmentStarted,
} from '../occurrence/utils';
import { ENROLMENT_URL_PARAMS, ENROLMENT_ERRORS } from './constants';
import EnrolmentForm, {
  defaultInitialValues,
  EnrolmentFormFields,
} from './enrolmentForm/EnrolmentForm';
import styles from './enrolmentPage.module.scss';
import EventInfo from './eventInfo/EventInfo';
import OccurrenceTable from './occurrenceTable/OccurrenceTable';
import { getEnrolmentPayload, getCAPTCHAToken } from './utils';

const CreateEnrolmentPage: React.FC = () => {
  const { t } = useTranslation();
  const locale = useLocale();
  const {
    query: { eventId, occurrences },
  } = useRouter();

  const { data: eventData, loading } = useEventQuery({
    variables: { id: eventId as string, include: ['location'] },
  });

  const [enrolOccurrence] = useEnrolOccurrenceMutation();

  const goToEventPage = () => {
    Router.push(ROUTES.EVENT_DETAILS.replace(':id', eventId as string));
  };

  const event = eventData?.event;
  const {
    locationId: eventLocationId,
    neededOccurrences,
    occurrences: allOccurrences,
  } = getEventFields(event, locale);

  const selectedOccurrences = Array.isArray(occurrences)
    ? occurrences
    : [occurrences];

  const filteredOccurrences = (allOccurrences || []).filter((o) =>
    selectedOccurrences.includes(o.id)
  );

  const initialValues = React.useMemo(
    () => ({
      ...defaultInitialValues,
      minGroupSize: Math.max(
        ...filteredOccurrences.map((item) => item.minGroupSize)
      ),
      maxGroupSize: Math.min(
        ...filteredOccurrences.map((item) =>
          Math.min(
            item.maxGroupSize,
            item.amountOfSeats - (item.seatsTaken || 0)
          )
        )
      ),
      language: locale.toUpperCase(),
    }),
    [filteredOccurrences, locale]
  );

  const submit = async (values: EnrolmentFormFields) => {
    try {
      const token = await getCAPTCHAToken();
      const data = await enrolOccurrence({
        variables: {
          input: {
            ...getEnrolmentPayload({
              occurrenceIds: filteredOccurrences.map((item) => item.id),
              values,
            }),
            captchaKey: token,
          },
        },
      });
      Router.push({
        pathname: ROUTES.EVENT_DETAILS.replace(':id', eventId as string),
        query: {
          ...omit(Router.query, [
            ENROLMENT_URL_PARAMS.EVENT_ID,
            ENROLMENT_URL_PARAMS.OCCURRENCES,
          ]),
          [ENROLMENT_URL_PARAMS.NOTIFICATION_TYPE]:
            data.data?.enrolOccurrence?.enrolments?.[0]?.notificationType,
          [ENROLMENT_URL_PARAMS.ENROLMENT_CREATED]: true,
        },
      });
    } catch (e) {
      toast(t('enrolment:errors.createFailed'), {
        type: toast.TYPE.ERROR,
      });
    }
  };

  const error = React.useMemo(() => {
    if (!isEnrolmentStarted(event))
      return ENROLMENT_ERRORS.ENROLMENT_NOT_STARTED_ERROR;
    if (filteredOccurrences.some((o) => isEnrolmentClosed(o, event)))
      return ENROLMENT_ERRORS.ENROLMENT_CLOSED_ERROR;
    if (filteredOccurrences.some((o) => !hasOccurrenceSpace(o)))
      return ENROLMENT_ERRORS.NOT_ENOUGH_CAPACITY_ERROR;
    if (neededOccurrences !== filteredOccurrences.length)
      return ENROLMENT_ERRORS.INVALID_OCCURRENCE_AMOUNT_ERROR;
    return null;
  }, [event, filteredOccurrences, neededOccurrences]);

  return (
    <PageWrapper title={t('enrolment:pageTitle')}>
      <LoadingSpinner isLoading={loading}>
        {event ? (
          <div className={styles.enrolmentPage}>
            <Container>
              <div className={styles.buttonWrapper}>
                <Button
                  iconLeft={<IconArrowLeft />}
                  onClick={goToEventPage}
                  variant="secondary"
                >
                  {t('enrolment:buttonBack')}
                </Button>
              </div>

              <h1>{t('enrolment:title')}</h1>
              <div className={styles.divider} />

              <EventInfo event={event} />
              {error ? (
                <Notification
                  labelText={translateValue(
                    'enrolment:errors.label.',
                    error,
                    t
                  )}
                  type="error"
                >
                  {translateValue('enrolment:errors.text.', error, t, {
                    count: neededOccurrences,
                  })}
                </Notification>
              ) : (
                <>
                  <OccurrenceTable
                    eventLocationId={eventLocationId || ''}
                    occurrences={filteredOccurrences}
                  />
                  <EnrolmentForm
                    initialValues={initialValues}
                    onSubmit={submit}
                  />
                </>
              )}
            </Container>
          </div>
        ) : (
          <NotFoundPage />
        )}
      </LoadingSpinner>
    </PageWrapper>
  );
};

export default CreateEnrolmentPage;
