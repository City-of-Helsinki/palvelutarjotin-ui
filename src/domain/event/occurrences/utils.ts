import {
  EventFieldsFragment,
  OccurrenceFieldsFragment,
} from '../../../generated/graphql';
import { ENROLMENT_ERRORS } from '../../enrolment/constants';
import {
  hasOccurrenceSpace,
  isEnrolmentClosed,
  isOccurrenceCancelled,
} from '../../occurrence/utils';

export function getEnrolmentError(
  occurrence: OccurrenceFieldsFragment,
  event: EventFieldsFragment
): ENROLMENT_ERRORS | null {
  if (isEnrolmentClosed(occurrence, event))
    return ENROLMENT_ERRORS.ENROLMENT_CLOSED_ERROR;
  if (!hasOccurrenceSpace(occurrence))
    return ENROLMENT_ERRORS.NOT_ENOUGH_CAPACITY_ERROR;
  if (isOccurrenceCancelled(occurrence)) {
    return ENROLMENT_ERRORS.ENROLMENT_CANCDELLED_ERROR;
  }

  return null;
}

export function getOrderedLanguages(
  languages: { id: string; name: string }[]
): string[] {
  const languageOrder = ['fi', 'sv', 'en'];
  const languageStrings = languages.map((l) => l.id.split('_')[0]);

  // order languages based on languageOrder, other languages go
  // to the end of the array
  return languageStrings.sort((a, b) => {
    const aIndex = languageOrder.indexOf(a);
    const bIndex = languageOrder.indexOf(b);

    if (aIndex > -1 && bIndex > -1) {
      return aIndex - bIndex;
    }

    if (aIndex > -1 && bIndex === -1) {
      return -1;
    }

    if (bIndex > -1 && aIndex === -1) {
      return 1;
    }

    return 0;
  });
}
