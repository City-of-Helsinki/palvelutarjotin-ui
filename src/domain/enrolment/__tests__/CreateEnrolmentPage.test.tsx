import { MockedResponse } from '@apollo/react-testing';
import { axe } from 'jest-axe';
import { advanceTo } from 'jest-date-mock';
import React from 'react';
import wait from 'waait';

import {
  Event,
  EventDocument,
  OccurrenceNode,
  PlaceDocument,
  StudyLevelsDocument,
} from '../../../generated/graphql';
import * as graphqlFns from '../../../generated/graphql';
import {
  fakePEvent,
  fakeLocalizedObject,
  fakeEvent,
  fakeOccurrences,
  fakePlace,
  fakeStudyLevels,
} from '../../../utils/mockDataUtils';
import {
  render,
  screen,
  act,
  waitFor,
  userEvent,
  configure,
  within,
} from '../../../utils/testUtils';
import CreateEnrolmentPage from '../CreateEnrolmentPage';
import * as utils from '../utils';

const eventId = '123321';
const occurrenceId1 = '123321';
const occurrenceId2 = '321123';
const occurrenceIds = [occurrenceId1, occurrenceId2];
const eventName = 'Testitapahtuma';
const locationId = 'locationId';

configure({ defaultHidden: true });
// eslint-disable-next-line @typescript-eslint/no-explicit-any
(utils.getCAPTCHAToken as any) = jest.fn().mockResolvedValue('captcha-token');

const mockBase = (event: Event): MockedResponse => ({
  request: {
    query: EventDocument,
    variables: { id: eventId, include: ['location'] },
  },
  result: {
    data: {
      event,
    },
  },
});

const mockStudyLevels = (): MockedResponse => ({
  request: {
    query: StudyLevelsDocument,
    variables: {},
  },
  result: {
    data: {
      studyLevels: fakeStudyLevels(),
    },
  },
});

// mock that has enrol time ended
const occurrenceOverrides1: Partial<OccurrenceNode>[] = [
  { id: occurrenceId1, startTime: new Date(2020, 8, 10, 12, 30) },
  { id: occurrenceId2, startTime: new Date(2020, 8, 10, 12, 30) },
  { id: 'sdfgdfhbf345345yreg' },
];

// mock that has enrol time ended
const mock1: MockedResponse[] = [
  mockBase(
    fakeEvent({
      name: fakeLocalizedObject(eventName),
      pEvent: fakePEvent({
        enrolmentStart: new Date(2020, 8, 7).toISOString(),
        enrolmentEndDays: 3,
        occurrences: fakeOccurrences(3, occurrenceOverrides1),
        neededOccurrences: 2,
      }),
    })
  ),
  mockStudyLevels(),
];

const occurrenceOverrides2: Partial<OccurrenceNode>[] = [
  { id: occurrenceId1, startTime: new Date(2020, 8, 25, 12, 30) },
  { id: occurrenceId2, startTime: new Date(2020, 8, 25, 12, 30) },
  { id: 'sdfgdfhbf345345yreg' },
];

const mock2: MockedResponse[] = [
  mockBase(
    fakeEvent({
      name: fakeLocalizedObject(eventName),
      pEvent: fakePEvent({
        enrolmentStart: new Date(2020, 8, 20).toISOString(),
        enrolmentEndDays: 3,
        occurrences: fakeOccurrences(3, occurrenceOverrides2),
        neededOccurrences: 2,
        mandatoryAdditionalInformation: true,
      }),
    })
  ),
  mockStudyLevels(),
];

const occurrenceOverrides3: Partial<OccurrenceNode>[] = [
  { id: occurrenceId1, startTime: new Date(2020, 8, 25, 12, 30) },
  { id: occurrenceId2, startTime: new Date(2020, 8, 26, 13, 20) },
  { id: 'sdfgdfhbf345345yreg' },
];

const mock3: MockedResponse[] = [
  {
    request: {
      query: PlaceDocument,
      variables: {
        id: locationId,
      },
    },
    result: {
      data: {
        place: fakePlace({ name: fakeLocalizedObject('Kirjasto') }),
      },
    },
  },
  mockBase(
    fakeEvent({
      name: fakeLocalizedObject(eventName),
      location: fakePlace({ id: locationId }),
      pEvent: fakePEvent({
        enrolmentStart: new Date(2020, 8, 7).toISOString(),
        enrolmentEndDays: 3,
        occurrences: fakeOccurrences(3, occurrenceOverrides3),
        neededOccurrences: 2,
      }),
    })
  ),
  mockStudyLevels(),
];

const mock4: MockedResponse[] = [
  mockBase(
    fakeEvent({
      name: fakeLocalizedObject(eventName),
      location: fakePlace({ id: locationId }),
      pEvent: fakePEvent({
        enrolmentStart: new Date(2020, 8, 7).toISOString(),
        enrolmentEndDays: 3,
        occurrences: fakeOccurrences(3, occurrenceOverrides3),
        neededOccurrences: 2,
        mandatoryAdditionalInformation: true,
      }),
    })
  ),
];

advanceTo(new Date(2020, 8, 8));
jest.setTimeout(20000);

// Notification component has a problem:
// "svg elements with an img role have an alternative text (svg-img-alt)"
test.skip('page is accessible', async () => {
  const { container } = render(<CreateEnrolmentPage />, {
    mocks: mock1,
    query: { eventId: eventId, occurrences: occurrenceIds },
  });

  await act(wait);
  expect(await axe(container)).toHaveNoViolations();
});

test('renders enrolment has ended text', async () => {
  render(<CreateEnrolmentPage />, {
    mocks: mock1,
    query: { eventId: eventId, occurrences: occurrenceIds },
  });

  await act(wait);

  expect(
    screen.getByRole('heading', { name: 'Ilmoittaudu tapahtumaan' })
  ).toBeInTheDocument();

  expect(screen.queryByText(eventName)).toBeInTheDocument();
  expect(screen.queryByText('Ilmoittautuminen päättynyt')).toBeInTheDocument();
  expect(
    screen.queryByText('Tapahtuman ilmoittautuminen on päättynyt')
  ).toBeInTheDocument();

  expect(
    screen.queryByRole('heading', { name: /ilmoittajan tiedot/i })
  ).not.toBeInTheDocument();
});

test('renders enrolment has not started yet text', async () => {
  render(<CreateEnrolmentPage />, {
    mocks: mock2,
    query: { eventId: eventId, occurrences: occurrenceIds },
  });

  await act(wait);

  expect(
    screen.getByRole('heading', { name: 'Ilmoittaudu tapahtumaan' })
  ).toBeInTheDocument();

  expect(screen.queryByText(eventName)).toBeInTheDocument();
  expect(
    screen.queryByText('Ilmoitautuminen ei ole avautunut')
  ).toBeInTheDocument();
  expect(
    screen.queryByText('Tapahtuman ilmoittautuminen ei ole vielä avautunut')
  ).toBeInTheDocument();

  expect(
    screen.queryByRole('heading', { name: /ilmoittajan tiedot/i })
  ).not.toBeInTheDocument();
});

test('renders form and user can fill it and submit', async () => {
  // eslint-disable-next-line import/namespace
  const enrolOccurrenceMock = jest.fn();
  jest
    .spyOn(graphqlFns, 'useEnrolOccurrenceMutation')
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    .mockReturnValue([enrolOccurrenceMock] as any);
  render(<CreateEnrolmentPage />, {
    mocks: mock3,
    query: { eventId: eventId, occurrences: occurrenceIds },
  });

  await waitFor(() => {
    expect(
      screen.queryByRole('heading', { name: /ilmoittajan tiedot/i })
    ).toBeInTheDocument();
  });

  expect(
    screen.getByRole('heading', { name: 'Ilmoittaudu tapahtumaan' })
  ).toBeInTheDocument();

  expect(
    screen.getByRole('heading', { name: 'Valitut tapahtuma-ajat' })
  ).toBeInTheDocument();

  await waitFor(() => {
    expect(screen.queryAllByText('Kirjasto')).toHaveLength(2);
  });

  expect(
    screen.queryByRole('row', {
      name: '25.09.2020 12:30 – 12:30 Kirjasto 30 / 30',
    })
  ).toBeInTheDocument();

  expect(
    screen.queryByRole('row', {
      name: '26.09.2020 13:20 – 12:30 Kirjasto 30 / 30',
    })
  ).toBeInTheDocument();

  userEvent.type(
    screen.getByRole('textbox', { name: /nimi/i }),
    'Nimi Niminen'
  );
  userEvent.type(
    screen.getByRole('textbox', { name: /sähköpostiosoite/i }),
    'testi@testi.fi'
  );
  userEvent.type(
    screen.getByRole('textbox', { name: /puhelinnumero/i }),
    '123321123'
  );

  userEvent.type(
    screen.getByRole('textbox', { name: /puhelinnumero/i }),
    '123321123'
  );
  userEvent.type(
    screen.getByRole('textbox', { name: /päiväkoti \/ koulu \/ oppilaitos/i }),
    'Testikoulu'
  );
  userEvent.type(screen.getByRole('textbox', { name: /ryhmä/i }), '4a');

  // select grade from dropdown
  userEvent.click(
    screen.getByRole('button', { name: /luokka-aste valitse\.\.\./i })
  );

  userEvent.click(screen.getByRole('option', { name: /4\. luokka/i }));
  userEvent.click(screen.getByRole('option', { name: /2\. luokka/i }));

  // close dropdown
  userEvent.click(screen.getByRole('button', { name: /luokka-aste/i }));

  userEvent.type(screen.getByLabelText(/lapsia/i), '10');
  userEvent.type(screen.getByLabelText(/aikuisia/i), '2');

  expect(
    screen.getByRole('checkbox', { name: /sama kuin ilmoittaja/i })
  ).toBeChecked();
  userEvent.click(screen.getByRole('checkbox', { name: /sähköpostilla/i }));
  userEvent.click(screen.getByText(/tekstiviestillä/i));

  userEvent.click(screen.getByRole('button', { name: /Suomi/ }));
  userEvent.click(screen.getByRole('option', { name: /suomi/i }));

  userEvent.type(
    screen.getByRole('textbox', { name: /lisätiedot \(valinnainen\)/i }),
    'Lisätietoja ilmoittautumiseen'
  );

  // test that error notification works
  userEvent.click(
    screen.getByRole('button', { name: /lähetä ilmoittautuminen/i })
  );
  const alertContainer = await screen.findByRole('alert');
  expect(alertContainer).toHaveTextContent(/Virhe lomakkeessa/i);
  expect(alertContainer).toHaveTextContent(
    /hyväksyn tietojeni jakamisen tapahtuman järjestäjän kanssa/i
  );

  userEvent.click(
    screen.getByRole('checkbox', {
      name: /hyväksyn tietojeni jakamisen tapahtuman järjestäjän kanssa/i,
    })
  );

  await waitFor(() => {
    expect(screen.queryByRole('alert')).not.toBeInTheDocument();
  });

  userEvent.click(
    screen.getByRole('button', { name: /lähetä ilmoittautuminen/i })
  );

  await waitFor(() => {
    expect(enrolOccurrenceMock).toHaveBeenCalledWith({
      variables: {
        input: {
          captchaKey: 'captcha-token',
          notificationType: 'EMAIL_SMS',
          occurrenceIds: ['123321', '321123'],
          person: undefined,
          studyGroup: {
            amountOfAdult: 2,
            extraNeeds: 'Lisätietoja ilmoittautumiseen',
            groupName: '4a',
            groupSize: 10,
            name: 'Testikoulu',
            person: {
              emailAddress: 'testi@testi.fi',
              language: 'FI',
              name: 'Nimi Niminen',
              phoneNumber: '123321123123321123',
            },
            studyLevels: ['GRADE_4', 'GRADE_2'],
          },
        },
      },
    });
  });
});

test('render and focuses error notification correctly', async () => {
  render(<CreateEnrolmentPage />, {
    mocks: mock3,
    query: { eventId: eventId, occurrences: occurrenceIds },
  });

  await waitFor(() => {
    expect(
      screen.queryByRole('heading', { name: /ilmoittajan tiedot/i })
    ).toBeInTheDocument();
  });

  userEvent.click(
    screen.getByRole('button', { name: /lähetä ilmoittautuminen/i })
  );

  const alertContainer = await screen.findByRole('alert');
  expect(alertContainer).toHaveTextContent(/Virhe lomakkeessa/i);

  const withinAlertContainer = within(alertContainer);

  const requiredFieldLabels = [
    'Hyväksyn tietojeni jakamisen tapahtuman järjestäjän kanssa',
    'Nimi',
    'Puhelinnumero',
    'Sähköpostiosoite',
    'Päiväkoti / koulu / oppilaitos',
    'Ryhmä',
    'Lapsia',
    'Aikuisia',
    'Luokka-aste',
  ];

  requiredFieldLabels.forEach((label) => {
    expect(withinAlertContainer.getByText(label)).toBeInTheDocument();
  });

  expect(alertContainer).toHaveFocus();

  userEvent.tab();

  expect(screen.getByRole('textbox', { name: 'Nimi' })).toHaveFocus();
});

test('amount of adults is limited to 1 when there is 1 seat left after children', async () => {
  render(<CreateEnrolmentPage />, {
    mocks: mock3,
    query: { eventId: eventId, occurrences: occurrenceIds },
  });
  await waitFor(() => {
    expect(screen.getByLabelText(/lapsia/i)).toBeInTheDocument();
  });
  userEvent.type(screen.getByLabelText(/lapsia/i), '19');
  userEvent.type(screen.getByLabelText(/aikuisia/i), '2');
  userEvent.tab();
  await waitFor(() => {
    expect(
      screen.getByText(
        /Arvon tulee olla enintään 1 yhdessä lasten lukumäärän kanssa/i
      )
    ).toBeInTheDocument();
  });
});

test('amount of adults is limited to 0 when children fills the group in whole', async () => {
  render(<CreateEnrolmentPage />, {
    mocks: mock3,
    query: { eventId: eventId, occurrences: occurrenceIds },
  });
  await waitFor(() => {
    expect(screen.getByLabelText(/lapsia/i)).toBeInTheDocument();
  });
  userEvent.type(screen.getByLabelText(/lapsia/i), '20');
  userEvent.type(screen.getByLabelText(/aikuisia/i), '2');
  userEvent.tab();
  await waitFor(() => {
    expect(
      screen.getByText(
        /Arvon tulee olla enintään 0 yhdessä lasten lukumäärän kanssa/i
      )
    ).toBeInTheDocument();
  });
});

test('adults should not validated against max group limit when there are already too many children', async () => {
  render(<CreateEnrolmentPage />, {
    mocks: mock3,
    query: { eventId: eventId, occurrences: occurrenceIds },
  });
  await waitFor(() => {
    expect(screen.getByLabelText(/lapsia/i)).toBeInTheDocument();
  });
  userEvent.type(screen.getByLabelText(/lapsia/i), '21');
  userEvent.type(screen.getByLabelText(/aikuisia/i), '2');
  userEvent.tab();
  await waitFor(() => {
    expect(
      screen.getByText(/Arvon tulee olla enintään 20/i)
    ).toBeInTheDocument();
  });
  await waitFor(() => {
    expect(
      screen.queryByText(
        /Arvon tulee olla enintään 0 yhdessä lasten lukumäärän kanssa/i
      )
    ).not.toBeInTheDocument();
  });
});

test('mandatory additional information forces extraNeeds field to be required', async () => {
  render(<CreateEnrolmentPage />, {
    mocks: mock4,
    query: { eventId: eventId, occurrences: occurrenceIds },
  });
  await waitFor(() => {
    expect(screen.getByLabelText(/Lisätiedot/i)).toBeInTheDocument();
  });
  userEvent.type(screen.getByRole('textbox', { name: /Lisätiedot/i }), '');
  userEvent.tab();
  await waitFor(() => {
    expect(screen.getByText(/Tämä kenttä on pakollinen/i)).toBeInTheDocument();
  });
});
