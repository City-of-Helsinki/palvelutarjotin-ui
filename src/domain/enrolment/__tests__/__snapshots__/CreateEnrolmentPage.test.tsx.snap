// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders form and user can fill it and submit and form is saved to local storage 1`] = `
Object {
  "dirty": true,
  "errors": Object {},
  "initialErrors": Object {},
  "initialTouched": Object {},
  "initialValues": Object {
    "hasEmailNotification": true,
    "hasSmsNotification": false,
    "isMandatoryAdditionalInformationRequired": false,
    "isSameResponsiblePerson": true,
    "isSharingDataAccepted": false,
    "language": "FI",
    "maxGroupSize": 20,
    "minGroupSize": 10,
    "person": Object {
      "emailAddress": "",
      "name": "",
      "phoneNumber": "",
    },
    "studyGroup": Object {
      "amountOfAdult": "",
      "extraNeeds": "",
      "groupName": "",
      "groupSize": "",
      "name": "",
      "person": Object {
        "emailAddress": "",
        "name": "",
        "phoneNumber": "",
      },
      "studyLevels": Array [],
    },
  },
  "isSubmitting": false,
  "isValid": true,
  "isValidating": false,
  "submitCount": 2,
  "touched": Object {
    "hasEmailNotification": true,
    "hasSmsNotification": true,
    "isMandatoryAdditionalInformationRequired": true,
    "isSameResponsiblePerson": true,
    "isSharingDataAccepted": true,
    "language": true,
    "maxGroupSize": true,
    "minGroupSize": true,
    "person": Object {
      "emailAddress": true,
      "name": true,
      "phoneNumber": true,
    },
    "studyGroup": Object {
      "amountOfAdult": true,
      "extraNeeds": true,
      "groupName": true,
      "groupSize": true,
      "name": true,
      "person": Object {
        "emailAddress": true,
        "name": true,
        "phoneNumber": true,
      },
      "studyLevels": Array [
        true,
        true,
      ],
    },
  },
  "validateOnBlur": true,
  "validateOnChange": true,
  "validateOnMount": false,
  "values": Object {
    "hasEmailNotification": true,
    "hasSmsNotification": true,
    "isMandatoryAdditionalInformationRequired": false,
    "isSameResponsiblePerson": true,
    "isSharingDataAccepted": true,
    "language": "FI",
    "maxGroupSize": 20,
    "minGroupSize": 10,
    "person": Object {
      "emailAddress": "",
      "name": "",
      "phoneNumber": "",
    },
    "studyGroup": Object {
      "amountOfAdult": 2,
      "extraNeeds": "Lis√§tietoja ilmoittautumiseen",
      "groupName": "4a",
      "groupSize": 10,
      "name": "Testikoulu",
      "person": Object {
        "emailAddress": "testi@testi.fi",
        "name": "Nimi Niminen",
        "phoneNumber": "123321123123321123",
      },
      "studyLevels": Array [
        "GRADE_4",
        "GRADE_2",
      ],
    },
  },
}
`;
